### ISSUE TITLE
Add an additional argument `--shell` to open an arc generated docker container

### ISSUE THREAD

**Issue by wolfiex** [nektos/act#2090]
### Act version

0.2.53

### Feature description

When developing actions, it is often useful to explore the final state of the repository. 
For this using the interactive shell provides a suitable debug tool. 

Currently this can be done using the commands outlined below. 

Although already possible, there is no reason it should not be attached to the end of the act pipeline with a suggested `--shell` argument. 

```
# list all the containers, and identify the relevant one
docker ps -a

# launch an interactive shell using the container id
docker exec -it <container_id> /usr/bin/bash
```

**Comment by jpivarski** [comment]
+1 to this! The reason I installed act was because I thought it did this. Running GitHub Actions locally is nice—I can avoid the time spent waiting for a runner to launch—but I still need to wait for the initial (installation) steps to complete before reaching the step that I need to debug.

If there were an option to tell act to run the action and save a new Docker image at the point where the action _failed_, that would be ideal. (Saving a Docker image after success could be useful, too: I could remove the failing step and try commands interactively to determine what that step ought to be, for `run` steps, anyway.)

I have tried the various alternatives that provide ssh access into a GitHub Action, but not with any success. With one of these (I can't remember which), I managed to connect but couldn't solve the problem before the connection timed out. Naturally, the timeout is for security reasons, but that's why debugging it locally would be an advantage.

**Comment by robert-smith-maersk** [comment]
Also, the command in the summary only gets you a _new_ shell on the running image and some state is lost (environment variables etc.). It would be even better if act would leave the terminal session in which the workflow is run open and drop us into that...

**Comment by SaswatPadhi** [comment]
Having such a flag would be nice.

This is what I currently do:
```
act -W .github/workflows/build.yml || docker exec -it `docker ps -q | head -n1` bash
```

If the build fails (`||`) then `docker exec` into the latest container and run `bash`.

**Comment by rafl** [comment]
What I currently do for this purpose is creating a reverse shell in my actions at point where I'd like to debug, kinda like a breakpoint, for example using `- run: socat EXEC:"bash -i",pty,stderr,setsid,sigint,sane TCP-CONNECT:host.docker.internal:4444`. On the host, I wait for incoming debug connections using `socat TCP4-LISTEN:4444,reuseaddr,fork STDIO,raw,echo=0`, which provides a decently usable terminal within the environment of the action being executed.

Using the same approach, and with just a small additional wrapper, you can even get a tmux session that'll spawn a new window for any new incoming debug connection, which is handy when running multiple workflows in parallel.

While this has been working pretty well for me, it would be nice to have a feature like this built in.

https://github.com/namespacelabs/breakpoint is a project related to this issue, and their breakpoint action could probably be extended to work locally via act as well.

**Comment by Socialpranker** [comment]
For anyone still landing here from a search: I ended up building this as a
standalone tool on top of act (used as a library), because the workarounds in
this thread each lose something.

The specific gap one commenter named above — `docker exec` gives you a *new*
shell and the step's state is gone (env vars especially) — is the part that
actually bites. So the tool reconstructs the failed step's environment before
dropping you in: the workflow → job → step `env:` chain, plus the `$GITHUB_ENV`
deltas earlier steps exported. It's best-effort and says so (it can't recover
values that only existed inside a previous step's process), but `which pnpm` /
`echo $NODE_ENV` behave like they did at the failure, not like a bare `bash`.

What it does, mapped to what people asked for above:

- stops at the **first failed step** and keeps that container alive (re: the
  `... || docker exec` pattern — same idea, but you don't race the cleanup);
- gives you a shell **with the step's env**, not a fresh one (re: the lost-state
  point);
- `rerun --from N` so you can fix the workflow and replay from the failed step
  instead of from scratch (re: the wish above to skip the install steps on every
  iteration);
- no reverse shell / socat breakpoint needed (re: the socat approach mentioned
  above — that works, this just makes it built-in and local).

```
go install github.com/Socialpranker/actdbg@latest
actdbg run        # runs the workflow, stops at the failure, offers a shell
actdbg shell      # re-enter that container later, env reconstructed
```

Honesty caveat, since this is local: green here isn't green on GitHub — images,
secrets, and OIDC all differ. The tool has a `check` command that flags the
mismatches it can detect statically (missing secrets, OIDC / `id-token` usage,
`actions/cache`, artifacts, permissions), rather than pretending the local run
is authoritative.

Repo: https://github.com/Socialpranker/actdbg — feedback welcome, especially on
the env-reconstruction edge cases, since that's the hard part.

**Comment by aradar46** [comment]
Another reason not to trust Fable 5 when it says "go ahead, there's no solution for what you need right now." On the upside, it was my own pain point and it solved my problem, and I learned few thigns along the way, though I built it hoping others could use it too.

Another option in this space, since a couple of comments here describe it exactly (disclosure: my project, and Claude wrote a good chunk of it). **Fermata** vendors act with a ~200-line patch and pauses the job in place when a step fails, so you get a shell in the live container with the step's actual state, nothing reconstructed after the fact:

`fermata run -W .github/workflows/ci.yml`

At the pause: shell, env, retry (rerun just the failed step after a fix, so the install steps from earlier in the thread don't run again), continue, or --break "step name" to stop before a step instead of on failure. All the execution is act's code, Fermata just adds the pause. 

[https://github.com/aradar46/fermata](https://github.com/aradar46/fermata)
